File: go/hamming/hamming.go
package hamming

import ( "errors" )

func Distance(a, b string) (int, error) {
    // check if strings are equivalent
    if len(a) != len(b) {
        return -1, errors.New("Sequences are not of equal length. The Distance function only works on sequences of the same length.")
    }

    // count the differences
    count := 0
    for i := 0; i < len(a); i++ {
        if a[i] != b[i] {
            count += 1
        }
    }

    return count, nil
}

File: cpp/hamming/hamming.cpp
// hamming.cpp

#include <string>

using namespace std;

namespace hamming {

    int compute(string first, string second) {

        // first check if string lengths are equal
        if (first.length() != second.length()) {
            throw domain_error("DNA Sequences of unequal length cannot be used to calculate a Hamming distance.");
        }

        int distance = 0;

        // compare strings
        for (string::size_type index = 0; index < first.length(); index++) {
            if (first[index] != second[index]) {
                distance += 1;
            }
        }

        return distance;
    }

}